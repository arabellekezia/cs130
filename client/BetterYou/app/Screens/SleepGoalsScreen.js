import React from "react";
import { SafeAreaView, StyleSheet, View, Text, Linking } from "react-native";
import AppText from "../components/AppText";
import AppTextInput from "../components/AppTextInput";
import TextButton from "../components/TextButton";
import TitleText from "../components/TitleText";
import GoalsService from "../services/GoalsService"; 

const MIN_SLEEP_TIME = 0;

function SleepGoalsScreen() {
  const [sleepHours, setSleepHours] = React.useState(0);
  const [err, setError] = React.useState({ sleepHours: false });

  async function save() {
    const validSleepHours =
      sleepHours > MIN_SLEEP_TIME &&
      sleepHours <= 24 &&
      Boolean(Number(sleepHours));
    if (!validSleepHours) {
      setError({ sleepHours: true });
      return;
    }
    await GoalsService.setSleepDurationGoal(sleepHours);
  }

  function resetErrors() {
    setError({ sleepHours: false });
  }

  function displayErrorMessage(err) {
    if (err.sleepHours) {
      if (sleepHours <= MIN_SLEEP_TIME) {
        return (
          <ErrorMessage message="Your goal must be greater than 0 hours per night." />
        );
      }

      if (sleepHours > 24) {
        return (
          <ErrorMessage message="Your goal must be less than 24 hours per night." />
        );
      }
      if (!Boolean(Number(sleepHours))) {
        return <ErrorMessage message="Your goal must be a number." />;
      }
    }
  }

  return (
    <SafeAreaView style={styles.container}>
      <TitleText style={styles.header} children="Sleep Goals" />
      <View style={styles.form}>
        <AppText
          style={styles.question}
          children="What is your target number of hours of sleep per night?"
        />
        <AppTextInput
          style={styles.textInput}
          placeholder="7 hours"
          isError={err.sleepHours}
          keyboardType="numeric"
          onChangeText={(sleepHours) => {
            setSleepHours(sleepHours);
            resetErrors();
          }}
        />
        {displayErrorMessage(err)}
        <HealthInfo style={styles.healthInfoContainer} />
      </View>
      <TextButton
        style={styles.saveButton}
        name="Save"
        onPress={async () => {
          await save();
        }}
      />
    </SafeAreaView>
  );
}

function ErrorMessage({ message }) {
  return (
    <AppText
      style={{
        alignSelf: "flex-start",
        marginLeft: "2.5%",
        color: "red",
      }}
      children={message}
    />
  );
}

function HealthInfo({ style }) {
  const sleepDurationLink =
    "https://www.cdc.gov/sleep/about_sleep/how_much_sleep.html";
  return (
    <View style={style}>
      <Text style={styles.informationalTextHeader}>
        Recommended sleep duration
      </Text>
      <Text style={styles.informationalText}>
        <Text style={styles.informationalText}>
          <Text>
            The amount of sleep needed is different for everybody. The CDC
            recommends that adults aim for
          </Text>
          <Text style={{ fontWeight: "bold" }}>{" 7 or more hours "}</Text>
          per night.
        </Text>
      </Text>
      <Text
        style={styles.learnMoreText}
        onPress={() => {
          Linking.openURL(sleepDurationLink);
        }}
      >
        {"Learn more"}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "white",
    alignItems: "center",
    justifyContent: "center",
  },
  header: {
    alignSelf: "flex-start",
    marginTop: "15%",
    marginLeft: "5%",
  },
  form: {
    flex: 1,
    width: "90%",
    alignItems: "center",
    flexDirection: "column",
  },
  question: {
    alignSelf: "flex-start",
    marginTop: "10%",
    marginBottom: "2%",
    color: "#474747",
    fontSize: 18,
    fontWeight: "bold",
  },
  healthInfoContainer: {
    marginTop: "10%",
    padding: "5%",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#eafbff",
    borderRadius: 24,
  },
  informationalTextHeader: {
    fontSize: 18,
    color: "#474747",
    textAlign: "center",
    fontWeight: "bold",
  },
  informationalText: {
    fontSize: 16,
    color: "#474747",
    textAlign: "left",
    marginTop: "5%",
    lineHeight: 26,
  },
  learnMoreText: {
    fontSize: 16,
    color: "#0079d3",
    textAlign: "center",
    marginTop: "5%",
    textDecorationLine: "underline",
  },
  saveButton: {
    alignSelf: "flex-end",
    marginBottom: "10%",
    minWidth: "90%",
  },
});

export default SleepGoalsScreen;
