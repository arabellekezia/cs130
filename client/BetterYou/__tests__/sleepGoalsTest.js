import {render, cleanup, fireEvent} from '@testing-library/react-native';
import React from 'react';
import SleepGoalsScreen from "../app/Screens/SleepGoalsScreen"

describe('Sleep goals input form', () => {
  let wrapper;
  beforeEach(() => {
      wrapper = render(<SleepGoalsScreen />);
    });
  afterEach(() => {
      cleanup();
      wrapper = null;
    });
  it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
  });

  it("Shows error message if target sleep is less than 24 hours per night", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("7 hours"); 
    const enteredGoal = "25";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Your goal must be less than 24 hours per night.");
  });

  it("Shows error message if target sleep is less than 0 hours per night", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("7 hours"); 
    const enteredGoal = "-10";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Your goal must be greater than 0 hours per night.");
  });

  it("Shows error message if the input is not a number.", () => {
    const {getByText, queryByTestId, getByPlaceholderText} = wrapper;

    const goalTextInput = getByPlaceholderText("7 hours"); 
    const enteredGoal = "6,";
    fireEvent(goalTextInput, 'onChangeText', enteredGoal);
    
    const saveButton = getByText("Save");
    fireEvent.press(saveButton);
  
    const validationError = queryByTestId("error-message");
    expect(validationError).toBeTruthy();
    
    expect(validationError.props.children)
        .toBe("Your goal must be a number.");
  });
});

