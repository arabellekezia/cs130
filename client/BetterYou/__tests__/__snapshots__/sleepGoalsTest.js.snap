// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sleep goals input form renders correctly 1`] = `
<RCTScrollView
  collapsable={false}
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  keyboardShouldPersistTaps="handled"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 30,
              "fontWeight": "bold",
            },
            Object {
              "alignSelf": "flex-start",
              "marginLeft": "5%",
              "marginTop": "15%",
            },
          ]
        }
      >
        Sleep Goals
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "width": "90%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {},
              Object {
                "alignSelf": "flex-start",
                "color": "#474747",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": "2%",
                "marginTop": "10%",
              },
            ]
          }
        >
          What is your target number of hours of sleep per night?
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#efefef",
              "borderRadius": 10,
              "flexDirection": "row",
              "marginVertical": 10,
              "padding": 15,
              "width": "100%",
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            keyboardType="numeric"
            maxLength={25}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="7 hours"
            rejectResponderTermination={true}
            style={
              Object {
                "fontSize": 18,
              }
            }
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#eafbff",
              "borderRadius": 24,
              "justifyContent": "center",
              "marginTop": "10%",
              "padding": "5%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#474747",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Recommended sleep duration
          </Text>
          <Text
            style={
              Object {
                "color": "#474747",
                "fontSize": 16,
                "lineHeight": 26,
                "marginTop": "5%",
                "textAlign": "left",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#474747",
                  "fontSize": 16,
                  "lineHeight": 26,
                  "marginTop": "5%",
                  "textAlign": "left",
                }
              }
            >
              <Text>
                The amount of sleep needed is different for everybody. The CDC recommends that adults aim for
              </Text>
              <Text
                style={
                  Object {
                    "fontWeight": "bold",
                  }
                }
              >
                 7 or more hours 
              </Text>
              per night.
            </Text>
          </Text>
          <Text
            onPress={[Function]}
            style={
              Object {
                "color": "#0079d3",
                "fontSize": 16,
                "marginTop": "5%",
                "textAlign": "center",
                "textDecorationLine": "underline",
              }
            }
          >
            Learn more
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "#007AFF",
              "borderRadius": 10,
              "justifyContent": "center",
              "marginBottom": "10%",
              "minWidth": "90%",
              "paddingHorizontal": 30,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            style={
              Array [
                Object {},
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "600",
                },
              ]
            }
          >
            Save
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
